src/odemis/acq/align/autofocus.py

```diff
@@ -756,6 +756,135 @@ def Sparc2AutoFocus(align_mode, opm, streams=None, start_autofocus=True):
+    """
+    Provides the ability to check the progress of the Sparc2 manual focus
+    procedure in a Future or even cancel it.
+        When the manual focus button is toggled:
+            - Turn on the light
+            - Change the optical path (closing the slit)
+            - Activate the first focus detector stream (if any)
+        When the manual focus button is untoggled:
+            - Change the optical path (opening the slit)
+            - Turn off the light
+       turn_on (bool): Toggle the manual focus button on/off
+       align_mode (str): OPM mode, spec-focus or spec-fiber-focus, streak-focus
+       opm: OpticalPathManager object
+       bl: bright light object
+       streams: list of streams
+       return (ProgressiveFuture -> for the _DoSparc2ManualFocus function)
+       raises:
+               CancelledError if cancelled
+               LookupError if procedure failed
+       """
+    est_start = time.time() + 0.1
+    if turn_on:
+        manual_focus_loading_times = (10, 1)
+    else:
+        manual_focus_loading_times = (1, 10)
+    f = model.ProgressiveFuture(start=est_start, end=est_start + sum(manual_focus_loading_times))
+    f._manual_focus_state = RUNNING
+    # Time for each action left
+    f._actions_time = list(manual_focus_loading_times)
+    f.task_canceller = _CancelSparc2ManualFocus
+    f._manual_focus_lock = threading.Lock()
+    f._running_subf = model.InstantaneousFuture()
+    executeAsyncTask(f, _DoSparc2ManualFocus,
+                     args=(f, turn_on, opm, bl, align_mode))
+    return f
+
+def _DoSparc2ManualFocus(future, turn_on, opm, bl, align_mode):
+    """
+    The actual implementation of the manual focus procedure, run asynchronously
+    :param future: the future object that is used to represent the task
+    :param turn_on (bool): Toggle the manual focus button on/off
+    :param opm: OpticalPathManager object
+    :param bl: brightlight object
+    :param align_mode: OPM mode, spec-focus or spec-fiber-focus, streak-focus
+    """
+    try:
+        if turn_on:
+            # Manual focus button is toggled
+            # First close slit, then switch on calibration light
+            # Go to the special focus mode (=> close the slit)
+            if future._manual_focus_state == CANCELLED:
+                logging.info("Manual focus procedure cancelled before closing the slit")
+                raise CancelledError()
+
+            future._running_subf = opm.setPath(align_mode)
+            future._running_subf.result()
+            future._actions_time.pop(0)
+            future.set_progress(end=time.time() + sum(future._actions_time))
+
+            # Turn on the light
+            if future._manual_focus_state == CANCELLED:
+                logging.info("Manual focus procedure cancelled before the light is on")
+                raise CancelledError()
+            bl.power.value = bl.power.range[1]
+            bl.emissions.value = [1] * len(bl.emissions.value)
```

Comment:
Actually, you don't need to wait for the path to be ready before turning on the light. You can do it at the same time. So in-between `opm.setPath(align_mode)` and `.result()`. In this case, you only have one step (of 10s).

Reply:
Done with the suggested refactoring. Just kept the canceller so at least the progress bar appears stopped (even if the opm future didn't get cancelled).